### PR TITLE
add flash messages

### PR DESCRIPTION
**Pull request by DmitryS1989**
https://trello.com/c/xhYvYF1Y
card_created
g)
acc_updated
sign_in
sign_out

### CODE REVIEW

File: app/controllers/boards_controller.rb [hunk 1]
```diff
@@ -13,22 +13,29 @@ def new
   end
 
   def create
```
Comment (B0RlS) [inline app/controllers/boards_controller.rb:15]: Could you please add specs

Comment (DmitryS1989) [reply app/controllers/boards_controller.rb:15]: You asked for specs, I gave you specs)))

File: app/controllers/cards_controller.rb [hunk 2]
```diff
@@ -16,24 +16,28 @@ def new
...
     if @card.save
+      flash[:success] = 'Card created'
       respond_to { |format| format.html { redirect_to board_cards_path(@board) } }
     else
+      flash_error
       respond_to { |format| format.html { render :new, status: :unprocessable_entity } }
     end
   end
 
   def update
```
Comment (AnastasiyaKireyeva) [inline app/controllers/cards_controller.rb:27]: can you, please, change it a bit according to boards#update?

File: app/models/board.rb [hunk 3]
```diff
@@ -3,4 +3,6 @@
 class Board < ApplicationRecord
   belongs_to :user
   has_many :cards, dependent: :destroy
+  validates :name, presence: true
+  validates :description, presence: true
```
Comment (AnastasiyaKireyeva) [inline app/models/board.rb]: can you validate the length here, please?

File: spec/factories/board.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :board do
+    user_id { '1' }
```
Comment (AnastasiyaKireyeva) [inline spec/factories/board.rb]:
```suggestion
    user_id { create(:user).id }
```

File: spec/factories/card.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :card do
+    board_id { '1' }
```
Comment (AnastasiyaKireyeva) [inline spec/factories/card.rb]: previous suggested changes can be applied here

**Comment by AnastasiyaKireyeva** [comment]
spec for board model is missed here

File: spec/requests/boards_spec.rb [hunk 6]
```diff
@@ -1,40 +1,189 @@
...
   end
 
   describe 'GET /show' do
     context 'when the user is not logged in' do
-      it 'returns redirect' do
-        get '/boards/show'
-        expect(response).to have_http_status(:redirect)
+      before do
+        get "/boards/#{board.id}"
       end
```
Comment (AnastasiyaKireyeva) [inline spec/requests/boards_spec.rb]:
```suggestion
      before { get "/boards/#{board.id}" }
```

File: spec/requests/boards_spec.rb [hunk 7]
```diff
@@ -1,40 +1,189 @@
...
   end
 
   describe 'GET /edit' do
     context 'when the user is not logged in' do
-      it 'returns redirect' do
-        get '/boards/edit'
-        expect(response).to have_http_status(:redirect)
+      before do
+        get "/boards/#{board.id}/edit"
+      end
```
Comment (AnastasiyaKireyeva) [inline spec/requests/boards_spec.rb]:
```suggestion
      before { get "/boards/#{board.id}/edit" }
```

File: spec/requests/boards_spec.rb [hunk 8]
```diff
@@ -1,40 +1,189 @@
...
+
+      it { is_expected.to redirect_to(user_session_path) }
+      it { expect(flash[:alert]).to include('You need to sign in or sign up before continuing.') }
+    end
+
+    context 'when user loggged in' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/boards_spec.rb]:
```suggestion
    context 'when user is loggged in' do
```

File: spec/support/request_examples.rb [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  before { get url }
+
+  it { is_expected.to redirect_to(user_session_path) }
+  it { expect(flash[:alert]).to include('You need to sign in or sign up before continuing.') }
+end
+
+RSpec.shared_examples 'when the user is logged in' do |url, exceptions|
+  let(:user) { create :user }
+  let(:board) { create :board, user_id: user.id }
+  before do
```
Comment (AnastasiyaKireyeva) [inline spec/support/request_examples.rb:13]:
```suggestion
  let(:board) { create :board, user_id: user.id }
  
  before do
```

**Comment by AnastasiyaKireyeva** [comment]
you can combine all contexts when the user is logged in. in order not to do the same things several times

File: spec/requests/cards_spec.rb [hunk 10]
```diff
@@ -3,38 +3,228 @@
...
+
+      it { expect(response).to render_template(:edit) }
+      it { expect(flash[:error]).to include('Answer is required') }
+    end
+  end
+
+  describe 'DELETE /destroy' do
+    context 'when the user is not loggged in' do
+      it 'is expected to destroy the requested card' do
+        card1 = create(:card, board_id: board.id)
```
Comment (AnastasiyaKireyeva) [inline spec/requests/cards_spec.rb]: use `let`, please

File: spec/requests/boards_spec.rb [hunk 11]
```diff
@@ -1,40 +1,189 @@
...
+  describe 'DELETE /destroy' do
+    context 'when the user is not loggged in' do
+      it 'is expected to destroy the requested card' do
+        board1 = create(:board, user_id: user.id)
+        expect do
+          delete "/boards/#{board1.id}"
+        end.to change(Card, :count).by(0)
+      end
+
+      it 'is expected to include "Card deleted"' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/boards_spec.rb]: are you sure?

**Comment by AnastasiyaKireyeva** [comment]
1. please, use let + don't create unnecessary variables
2. make more global contexts, when it's possible
3. if there's a possibility to write a before block in 1 line, please, do it
4. reread your contexts and it blocks, please, many of them can be improved, both grammatically and in meaning

File: app/controllers/cards_controller.rb (hunk 2 above)
Comment (DmitryS1989) [reply app/controllers/cards_controller.rb:27]: I can't, I'm an artist that's how I see it

File: app/models/board.rb (hunk 3 above)
Comment (DmitryS1989) [reply app/models/board.rb]: I don’t think it’s in this task ... how did it even get here?)

File: spec/requests/cards_spec.rb (hunk 10 above)
Comment (DmitryS1989) [reply spec/requests/cards_spec.rb]: > spec for board model is missed here

I don’t think it’s in this task

**Comment by DmitryS1989** [comment]
> you can combine all contexts when the user is logged in. in order not to do the same things several time
What is the problem ? Is it unreadable? We test each controller method separately.

**Comment by DmitryS1989** [comment]
> 1. please, use let + don't create unnecessary variables
> 
>     2. make more global contexts, when it's possible
> 
>     3. if there's a possibility to write a before block in 1 line, please, do it
> 
>     4. reread your contexts and it blocks, please, many of them can be improved, both grammatically and in meaning

1.  ok
2. We test each controller method separately.
3. ok
4. I looked at the specs of another team. Are you prejudiced against us?)

File: spec/requests/cards_spec.rb [hunk 12]
```diff
@@ -188,7 +188,7 @@
   end
 
   describe 'DELETE /destroy' do
-    context 'when the user is not loggged in' do
+    context 'when the user is not logggd in' do
```
Comment (B0RlS) [inline spec/requests/cards_spec.rb]: logged*

**Comment by B0RlS** [comment]
Could you please fix the conflicts and use Present Simple in the commit names?

File: app/controllers/cards_controller.rb (hunk 2 above)
Comment (AnastasiyaKireyeva) [reply app/controllers/cards_controller.rb:27]: the comment is still relevant

**Comment by AnastasiyaKireyeva** [comment]
you misunderstood me about the structure of the tests. each method should be tested separately and divided into 2 main contexts 'when the user is logged in' and the opposite one. and other contexts should be inside them

File: app/controllers/cards_controller.rb (hunk 2 above)
Comment (AnastasiyaKireyeva) [reply app/controllers/cards_controller.rb:27]: the comment is STILL relevant

File: spec/requests/boards_spec.rb [hunk 13]
```diff
@@ -1,40 +1,149 @@
...
+    describe 'POST /create' do
+      before do
+        post '/boards', params: { board: { name: 'My Board', description: 'Test' } }
+        follow_redirect!
+      end
+
+      it { expect(response.body).to include('My Board') }
+      it { expect(flash[:success]).to include('Board created') }
+
+      context 'when name empty' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/boards_spec.rb]:
```suggestion
      context 'when name is empty' do
```

File: spec/requests/boards_spec.rb [hunk 14]
```diff
@@ -1,40 +1,149 @@
...
+        before { post '/boards', params: { board: { name: '', description: 'Test' } } }
+
+        it { expect(response).to render_template(:new) }
+        it { expect(flash[:error]).to include("Name can't be blank") }
+      end
+
+      context 'when description empty' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/boards_spec.rb]:
```suggestion
      context 'when description is empty' do
```

File: spec/requests/boards_spec.rb [hunk 15]
```diff
@@ -1,40 +1,149 @@
...
+      before do
+        patch board_url(board), params: { board: { name: 'My Board', description: 'Test' } }
+        follow_redirect!
+      end
+
+      it { expect(response.body).to include('My Board') }
+      it { expect(response.body).to include('Test') }
+      it { expect(flash[:success]).to include('Board updated') }
+
+      context 'when name empty' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/boards_spec.rb]:
```suggestion
      context 'when name is empty' do
```

File: spec/requests/boards_spec.rb [hunk 16]
```diff
@@ -1,40 +1,149 @@
...
+        before { patch board_url(board), params: { board: { name: '', description: 'Test' } } }
+
+        it { expect(response).to render_template(:edit) }
+        it { expect(flash[:error]).to include("Name can't be blank") }
+      end
+
+      context 'when description empty' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/boards_spec.rb]:
```suggestion
      context 'when description is empty' do
```

File: spec/requests/cards_spec.rb [hunk 17]
```diff
@@ -3,38 +3,174 @@
...
+    describe 'POST /create' do
+      before do
+        post "/boards/#{board.id}/cards/", params: correct_params
+        follow_redirect!
+      end
+
+      it { expect(response.body).to include('Where is my car dude?') }
+      it { expect(flash[:success]).to include('Card created') }
+
+      context 'when question empty' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/cards_spec.rb]:
```suggestion
      context 'when question is empty' do
```

File: spec/requests/cards_spec.rb [hunk 18]
```diff
@@ -3,38 +3,174 @@
...
+        before { post "/boards/#{board.id}/cards/", params: question_blank }
+
+        it { expect(response).to render_template(:new) }
+        it { expect(flash[:error]).to include('Question is required') }
+      end
+
+      context 'when answer empty' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/cards_spec.rb]:
```suggestion
      context 'when answer is empty' do
```

File: spec/requests/cards_spec.rb [hunk 19]
```diff
@@ -3,38 +3,174 @@
...
+        before { post "/boards/#{board.id}/cards/", params: answer_blank }
+
+        it { expect(response).to render_template(:new) }
+        it { expect(flash[:error]).to include('Answer is required') }
+      end
+
+      context 'when question so long' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/cards_spec.rb]:
```suggestion
      context 'when question is too long' do
```

File: spec/requests/cards_spec.rb [hunk 20]
```diff
@@ -3,38 +3,174 @@
...
+      before do
+        patch "/boards/#{board.id}/cards/#{card.id}", params: correct_params
+        follow_redirect!
+      end
+
+      it { expect(response.body).to include('Where is my car dude?') }
+      it { expect(response.body).to include('Gone in Sixty Seconds') }
+      it { expect(flash[:success]).to include('Card updated') }
+
+      context 'when question empty' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/cards_spec.rb]:
```suggestion
      context 'when question is empty' do
```

File: spec/requests/cards_spec.rb [hunk 21]
```diff
@@ -3,38 +3,174 @@
...
+        before { patch "/boards/#{board.id}/cards/#{card.id}", params: question_blank }
+
+        it { expect(response).to render_template(:edit) }
+        it { expect(flash[:error]).to include('Question is required') }
+      end
+
+      context 'when answer empty' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/cards_spec.rb]:
```suggestion
      context 'when answer is empty' do
```

File: spec/requests/cards_spec.rb [hunk 22]
```diff
@@ -3,38 +3,174 @@
...
+        before { patch "/boards/#{board.id}/cards/#{card.id}", params: answer_blank }
+
+        it { expect(response).to render_template(:edit) }
+        it { expect(flash[:error]).to include('Answer is required') }
+      end
+
+      context 'when question so long' do
```
Comment (AnastasiyaKireyeva) [inline spec/requests/cards_spec.rb]:
```suggestion
      context 'when question is too long' do
```

File: app/models/board.rb (hunk 3 above)
Comment (AnastasiyaKireyeva) [reply app/models/board.rb]: and this one is also relevant

File: app/controllers/cards_controller.rb (hunk 2 above)
Comment (B0RlS) [reply app/controllers/cards_controller.rb:27]: Anastasiya’s comment is still relevant

**Pull request merged**